### ISSUE TITLE
clarify CTE `[NOT] MATERIALIZED` options

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/docs#15419]
[Our docs on CTEs](https://www.cockroachlabs.com/docs/stable/common-table-expressions.html) currently describe the `MATERIALIZED` and `NOT MATERIALIZED` options as:

> Override the [optimizer](https://www.cockroachlabs.com/docs/v22.1/cost-based-optimizer)'s decision to materialize (i.e., store the results) of the common table expression. By default, the optimizer materializes the common table expression if it affects other objects in the database, or if it is used in the query multiple times.

This doesn't make the behavior of these queries clear. We should clarify the different between the three options. ("Parent query" refers to the statement that comes after the `WITH`, e.g., `SELECT * FROM t` in `WITH tmp AS (SELECT * FROM s) SELECT * FROM t`).

1. _Default_: The optimizer will inline the CTE into the parent query if it does not have side effects (i.e., it is not volatile [add link to volatility docs]) and the CTE is referenced only once in the parent query. 
2. `MATERIALIZED`: Forces the optimizer to never inline the CTE into the parent query.
3. `NOT MATERIALIZED`: Forces the optimizer to always inline the CTE into the parent query, even if it has side effects or is referenced multiple times. 

[Postgres's docs](https://www.postgresql.org/docs/14/queries-with.html#id-1.5.6.12.7) have a good description of the behavior, for reference:

> However, if a WITH query is non-recursive and side-effect-free (that is, it is a SELECT containing no volatile functions) then it can be folded into the parent query, allowing joint optimization of the two query levels. By default, this happens if the parent query references the WITH query just once, but not if it references the WITH query more than once. You can override that decision by specifying MATERIALIZED to force separate calculation of the WITH query, or by specifying NOT MATERIALIZED to force it to be merged into the parent query. The latter choice risks duplicate computation of the WITH query, but it can still give a net savings if each usage of the WITH query needs only a small part of the WITH query's full output.
